zebra-chain/src/addresses.rs

```diff
@@ -0,0 +1,222 @@
+    /// The testnet.
+    Testnet,
+}
+
+/// Transparent Zcash Addresses
+///
+/// In Bitcoin a single byte is used for the version field identifying
+/// the address type. In Zcash two bytes are used. For addresses on
+/// the production network, this and the encoded length cause the first
+/// two characters of the Base58Check encoding to be xed as “t3” for
+/// P2SH addresses, and as “t1” for P2PKH addresses. (This does not
+/// imply that a transparent Zcash address can be parsed identically
+/// to a Bitcoin address just by removing the “t”.)
+///
+/// https://zips.z.cash/protocol/protocol.pdf#transparentaddrencoding
+#[derive(Copy, Clone, Eq, PartialEq)]
+// #[cfg_attr(test, derive(Arbitrary))]
+pub enum TransparentAddress {
+    /// P2SH (Pay to Script Hash) addresses
+    PayToScriptHash {
+        /// Production, test, or other network
+        network: Network,
+        /// 20 bytes specifying a script hash.
+        script_hash: AddressPayloadHash,
+    },
+    /// P2PKH (Pay to Public Key Hash) addresses
+    PayToPublicKeyHash {
+        /// Production, test, or other network
+        network: Network,
+        /// 20 bytes specifying a public key hash, which is a RIPEMD-160
+        /// hash of a SHA-256 hash of a compressed ECDSA key encoding.
+        pub_key_hash: AddressPayloadHash,
+    },
+}
+
+impl From<Script> for TransparentAddress {
+    fn from(script: Script) -> Self {
+        TransparentAddress::PayToScriptHash {
+            network: Network::Mainnet,
+            script_hash: AddressPayloadHash::from(&script.0[..]),
+        }
+    }
+}
+
+impl From<PublicKey> for TransparentAddress {
+    fn from(pub_key: PublicKey) -> Self {
+        TransparentAddress::PayToPublicKeyHash {
+            network: Network::Mainnet,
+            pub_key_hash: AddressPayloadHash::from(&pub_key.serialize()[..]),
+        }
+    }
+}
+
+impl fmt::Debug for TransparentAddress {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut bytes = io::Cursor::new(Vec::new());
+        let _ = self.zcash_serialize(&mut bytes);
+
+        f.debug_tuple("TransparentAddress")
+            .field(&bs58::encode(bytes.get_ref()).with_check().into_string())
+            .finish()
+    }
+}
+
+impl ZcashSerialize for TransparentAddress {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        match self {
+            TransparentAddress::PayToScriptHash {
+                network,
+                script_hash,
+            } => {
+                if *network == Network::Mainnet {
+                    writer.write_all(&[0x1C, 0xBD][..])?
+                } else {
+                    // Dev network doesn't have a recommendation so we
+                    // default to testnet bytes.
+                    writer.write_all(&[0x1C, 0xBA][..])?
+                }
+                // XXX I'm asuming this is BigEndian because we're not
+                // explicitly making it LittleEndian?
```

Comment:
Unless Bitcoin does something very bizarre (real possibility! does it?) I don't think that endianness comes into play here, since it's reading and writing byte sequences rather than integers.